vm/cuttlefish/cuttlefish.go

```diff
@@ -139,7 +140,7 @@ func (inst *instance) Forward(port int) (string, error) {
 	}
 
 	cmd := fmt.Sprintf("nohup socat TCP-LISTEN:%d,fork TCP:%s", port, hostForward)
-	if err := inst.runOnHost(time.Second, cmd); err != nil && err != vmimpl.ErrTimeout {
+	if err := inst.runOnHost(time.Second, cmd); err != nil && !strings.Contains(err.Error(), "timedout after") {
```

Comment:
To be honest, I have mixed feelings about this line of code.

On the one hand, this looks very fragile to me.

It would be much better to refactor the way the `timeout after` error is returned from the [Run](https://github.com/eprucka3/syzkaller/blob/28cf466137d57c0f530aab91c12fe2a1e84bf0b4/pkg/osutil/osutil.go#L76) method - e.g. so that the special `timed out error` type is explicitly wrapped in the `VerboseError` type. When the `Run`'s code was written, there was just no [suitable error wrapping mechanism](https://go.dev/blog/go1.13-errors) in Go.

On the other hand, this is a quite large refactoring of the existing code. And it's [not](https://github.com/eprucka3/syzkaller/blob/28cf466137d57c0f530aab91c12fe2a1e84bf0b4/pkg/osutil/osutil.go#L126) just a simple replacement of `Title` with some nested `Error error`. So it's definitely an overkill in this case.

----

Let's try to look for alternatives instead.
I see you want to catch the `socat` execution errors early here and therefore give it some time to fail, then it's just left to itself.

* You can execute the program in background and then, after that 1 second, do one more execution to see if `socat` is still alive.
* You can write an a bit more sophisticated bash script so that the `socat` command is started, the script waits for some time and, if the command did not exit, just detaches from it.

In these cases you don't need to check whether the `runOnHost` finished with a timeout error. All errors are bad in these cases.

Reply:
The current code does not really catch those port forwarding errors -- even the ssh connection errors will be ignored. Also one extra ssh connection will be hanging for each VM instance.

If you don't really care about fetching `socat` execution errors (in the end, you'll later anyway notice if that part failed), it might be better to just run `nohup socat TCP-LISTEN:%d,fork TCP:%s &` using `runOnHost`.

Reply:
That's a good point; it will be caught when trying to forward the port, and return an error. I switched it back to runOnHost.

Reply:
Yes, now it's better!

And I would also leave the error handling code -- in the end, runOnHost can fail, too.

Reply:
Makes sense; thanks!

Reply:
Unfortunately, although the function runs in the background, the timeout causes the function to return an error even when working properly. I'll work on switching this to a bash script to fix the issue.

Reply:
This indeed hangs until the command has finished.
```
ssh localhost 'bash -c "sleep 10 &"'
```

And in this case it connects to the SSH and starts the command. We actually don't even need `&` when using `-f`.
```
ssh localhost -f 'bash -c "nohup sleep 10"'
```

Reply:
Thanks for the help! I tried running `ssh localhost -f` from runOnHost(), but the command hung, blocking future execution. When running the same command directly from os/exec, it does work correctly.

Reply:
Maybe the `-f` flag just did not make its way to the actual command when you ran it via `runOnHost()`